### PR TITLE
c_languages: Add CPPCheckBear

### PR DESCRIPTION
**Pull request by Redridge**
References https://github.com/coala-analyzer/coala-bears/issues/389

### CODE REVIEW

File: tests/c_languages/CPPCheckBearTest.py [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+}""".splitlines(keepends=True)
+
+
+CPPCheckEnableAllFalseTest = verify_local_bear(
+    CPPCheckBear, valid_files=(good_file,), invalid_files=(bad_file,),
+    settings={'enable_all_checks': 'False'})
+
+CPPCheckBearTest = verify_local_bear(CPPCheckBear,
+                                     valid_files=(),
+                                     invalid_files=(good_file, bad_file))
```
Comment (SanketDG) [inline tests/c_languages/CPPCheckBearTest.py]: could you maybe include a `very_good_file` that doesn't get picked up by CppCheckBear at all.

File: tests/c_languages/CPPCheckBearTest.py [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+    p[c] = 42; \
+}
+int main() {
+    f(100);
+    return 0;
+}""".splitlines(keepends=True)
+
+
+CPPCheckEnableAllFalseTest = verify_local_bear(
+    CPPCheckBear, valid_files=(good_file,), invalid_files=(bad_file,),
```
Comment (SanketDG) [inline tests/c_languages/CPPCheckBearTest.py]: nitpick: keep the indentation as same as lines 28-30

File: bears/c_languages/CPPCheckBear.py [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+    LANGUAGES = "C", "C++"
+
+    @staticmethod
+    def create_arguments(filename, file, config_file,
+                         enable_all_checks: bool=True):
+        """
+        :param enable_all_checks:
+            Report warnings and style issues too.
+        """
+        args = ("--template='{line}:{severity}:{id}:{message}'",)
```
Comment (SanketDG) [inline bears/c_languages/CPPCheckBear.py]: the quotes are not needed after `template=`

instead they are not escaped and the message appears with a trailing `'`.

```
bad.c
|   6| ••••f(100);
|    | [MAJOR] CPPCheckBear (arrayIndexOutOfBounds):
|    | Array 'p[10]' accessed at index 100, which is out of bounds.'
|    | The following actions are applicable to this result:
|    |  0: Apply no further actions.
|    |  1: Open the affected file(s) in an editor.
|    | Please enter the number of the action you want to execute (Ctrl-D to exit). Program terminated by user.
```

File: bears/c_languages/CPPCheckBear.py [hunk 4]
```diff
@@ -0,0 +1,35 @@
+from coalib.bearlib.abstractions.Linter import linter
+from coalib.results.RESULT_SEVERITY import RESULT_SEVERITY
+from coalib.settings.Setting import typed_list
+
+
+@linter(executable="cppcheck",
+        use_stdout=False,
```
Comment (SanketDG) [inline bears/c_languages/CPPCheckBear.py]: is this needed?

just doing `use_stderr=True` would work?

if there is any kind of useless output on stdout, you may ignore this.

**Comment by SanketDG** [comment]
@Redridge 

TODO change diff_message to result_message after https://github.com/coala-analyzer/coala-bears/pull/443 gets merged in.

Also, you use references in your commit message, any reason for that?

**Comment by Redridge** [comment]
@SanketDG I used "references" since it doesn't actually solve the issue. I will get going with the changes.

File: bears/c_languages/CPPCheckBear.py (hunk 4 above)
Comment (Redridge) [inline bears/c_languages/CPPCheckBear.py]: yes stdout is useless, it only prints irrelevant information

File: bears/c_languages/CPPCheckBear.py [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+    def create_arguments(filename, file, config_file,
+                         enable_all_checks: bool=True):
+        """
+        :param enable_all_checks:
+            Report warnings and style issues too.
+        """
+        args = ("--template={line}:{severity}:{id}:{message}",)
+
+        if enable_all_checks:
+            args += ('--enable=all',)
```
Comment (sils1297) [inline bears/c_languages/CPPCheckBear.py]: looks like you could also expose this in a way that a user can simply give you a list of checks to enable?

File: bears/c_languages/CPPCheckBear.py [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+    """
+    Report possible security weaknesses for C/C++.
+    For more information, consult <https://github.com/danmar/cppcheck>.
+    """
+
+    LANGUAGES = "C", "C++"
+
+    @staticmethod
+    def create_arguments(filename, file, config_file,
+                         enable_all_checks: bool=True):
```
Comment (sils1297) [inline bears/c_languages/CPPCheckBear.py]: maybe `enable_all_cpplint_checks`

File: bears/c_languages/CPPCheckBear.py (hunk 5 above)
Comment (Redridge) [inline bears/c_languages/CPPCheckBear.py]: All options can be found using cppcheck -h
but the problem I encounter is with the cppcheck versions. Basically different versions clasify the same issue with different severity (warn gets confused with style).
Also even though there are more options the error, warning, style, I couldn't find a piece of code that would trigger those.

Comment (Redridge) [inline bears/c_languages/CPPCheckBear.py]: The above described problem was the one that made my test fail. The cppcheck version on the CI is different for some reason.

File: bears/c_languages/CPPCheckBear.py [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+    def create_arguments(filename, file, config_file,
+                         enable: typed_list(str)=[]):
+        """
+        :param enable_all_checks:
+            Report warnings and style issues too.
+        """
+        args = ("--template={line}:{severity}:{id}:{message}",)
+
+        if enable:
+            args += ('--enable=' + "".join(","+str(s) for s in enable)[1:],)
```
Comment (SanketDG) [inline bears/c_languages/CPPCheckBear.py]: nitpick: there should be spaces around the second plus operator

**Comment by SanketDG** [comment]
ack 90d4581

File: bears/c_languages/CPPCheckBear.py [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+    """
+
+    LANGUAGES = "C", "C++"
+
+    @staticmethod
+    def create_arguments(filename, file, config_file,
+                         enable: typed_list(str)=[]):
+        """
+        :param enable_all_checks:
+            Report warnings and style issues too.
```
Comment (sils1297) [inline bears/c_languages/CPPCheckBear.py]: documentation wrong

**Comment by sils1297** [comment]
unack 90d4581

File: bears/c_languages/CPPCheckBear.py [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+        :param enable:
+            Choose specific issues to report. Issues that can be
+            reported are: all, warning, style, performance,
+            portability, information, unusedFunction,
+            missingInclude
+        """
+        args = ("--template={line}:{severity}:{id}:{message}",)
+
+        if enable:
+            args += ('--enable=' + "".join("," + str(s) for s in enable)[1:],)
```
Comment (sils1297) [inline bears/c_languages/CPPCheckBear.py]: why don't you simply do `','.join(enable)`?

**Comment by sils1297** [comment]
https://github.com/coala-analyzer/coala-bears/pull/448/commits/d90e05f6bc472f297e5edff7b3ab0b60e85b455b needs work

**Comment by sils1297** [comment]
and a rebase, then we can merge it :)

**Comment by sils1297** [comment]
ack https://github.com/coala-analyzer/coala-bears/pull/448/commits/7437fe725960053a946e1b274f8a32d17f7e10d9

**Comment by sils1297** [comment]
@rultor merge

**Comment by sils1297** [comment]
@rultor merge

**Comment by sils1297** [comment]
@Redridge awesome :) https://twitter.com/coala_analyzer/status/734715491370041344

**Pull request merged**